### DISCUSSION TITLE
Redis Adopts Dual Source-Available Licensing

### DISCUSSION THREAD

**Discussion by elf-pavlik** [taskforcesh/bullmq#2484] [General]
https://redis.com/blog/redis-adopts-dual-source-available-licensing/

> Beginning today, all future versions of Redis will be released with source-available licenses. Starting with Redis 7.4, Redis will be dual-licensed under the [Redis Source Available License (RSALv2)](https://redis.com/legal/rsalv2-agreement/) and [Server Side Public License (SSPLv1)](https://redis.com/legal/server-side-public-license-sspl/). Consequently, Redis will no longer be distributed under the three-clause Berkeley Software Distribution (BSD).

Is there a plan to switch to an open-source fork, for example, https://github.com/placeholderkv/placeholderkv

**Comment by manast** [comment]
A couple of things about this. First of all, BullMQ works perfectly with any Redis version from 6.2.0 and onwards. Basically there is no need to use any newer Redis version for using BullMQ. Secondly, the license, AFAIK but correct me if I am wrong, does only affect host providers, you should still be able to privately host it if you want, but most likely you are going to use any provider such as Redislabs, Memetria, AWS Elasticache, etc, etc, then you should not need to care either.

**Reply by elf-pavlik** [reply]
I maintain open-source software, which I would like people to easily host wherever they get their hosting from. It may be too early to see how cloud providers will respond, but I see it as something to watch. Whether some providers partner with Redis or others switch to alternatives like [KeyDB](https://github.com/Snapchat/KeyDB), [Garnet](https://github.com/microsoft/garnet), or a new fork of Redis, I would like to feel confident that my project having BullMQ as a dependency will still be easily deployable.

**Reply by manast** [reply]
BullMQ also is compatible with Dragonfly and Upstash.

**Reply by elf-pavlik** [reply]
Thank you. So far, I have found the following options:

* Open Source
  * [Valkey](https://www.linuxfoundation.org/press/linux-foundation-launches-open-source-valkey-community)
  * [KeyDB](https://github.com/Snapchat/KeyDB)
  * [Garnet](https://github.com/microsoft/garnet)
* Source Available
  * Redis
  * [Dragonfly](https://www.dragonflydb.io/)
* Closed Source
  * [Upstash](https://upstash.com/)

Is there a way to track compatibility with those products?

**Comment by HighDoping** [comment]
I've done some testing in https://github.com/misskey-dev/misskey/discussions/13612#discussioncomment-8878221. Seems BullMQ work well with KeyDB, but fails to work with Kvrocks or Garnet, maybe related to version check. Valkey don't seems to have a stable release yet.

**Reply by elf-pavlik** [reply]
A few days ago, I tried https://hub.docker.com/r/valkey/valkey, and it was erroring. I didn't have time to try to figure out what was causing the problem. Whenever I find time to dig into it I'll report here.

**Comment by HighDoping** [comment]
More testing using [#taskforcesh/bullmq-bench](https://github.com/taskforcesh/bullmq-bench) and Docker.

Test with latest Kvrocks(v2.8.0) produced following error.

    [1712840302590] BullMQ Benchmark: Running suite for: BullMQ 4.12.0
    [1712840302591] BullMQ Benchmark: Running benchmark: queue-add
    124 |         await RedisConnection.waitUntilReady(this._client);
    125 |         this.loadCommands();
    126 |         this.version = await this.getRedisVersion();
    127 |         if (this.skipVersionCheck !== true && !this.closing) {
    128 |             if ((0, utils_2.isRedisVersionLowerThan)(this.version, RedisConnection.minimumVersion)) {
    129 |                 throw new Error(`Redis version needs to be greater or equal than ${RedisConnection.minimumVersion} Current: ${this.version}`);
                                ^
    error: Redis version needs to be greater or equal than 5.0.0 Current: 4.0.0
        at /workspaces/Redisversion/node_modules/bullmq/dist/cjs/classes/redis-connection.js:129:23

Kvrocks only reports redis_version 4.0.0 so it's incompatible. It just added RESP3 support but still in development.

Test with latest Garnet(v1.0.4) produced following error.

    168 |     }
    169 |     shiftCommand(reply) {
    170 |         const item = this.redis.commandQueue.shift();
    171 |         if (!item) {
    172 |             const message = "Command queue state error. If you can reproduce this, please report it.";
    173 |             const error = new Error(message +
                                    ^
    error: Command queue state error. If you can reproduce this, please report it. Last error: ERR unknown command
        at shiftCommand (/workspaces/Redisversion/node_modules/ioredis/built/DataHandler.js:173:27)
        at returnError (/workspaces/Redisversion/node_modules/ioredis/built/DataHandler.js:33:22)
        at returnError (/workspaces/Redisversion/node_modules/ioredis/built/DataHandler.js:15:17)
        at execute (/workspaces/Redisversion/node_modules/redis-parser/lib/parser.js:542:9)
        at /workspaces/Redisversion/node_modules/ioredis/built/DataHandler.js:25:13
        at addChunk (node:stream:1958:43)
        at readableAddChunk (node:stream:1912:59)
        at data (node:net:52:17)

Seems to be compatibility issue with ioredis.

The migration mentioned in <https://github.com/taskforcesh/bullmq/issues/2490> may be needed to support these alternatives.

**Reply by manast** [reply]
The error with Garnet seems to be that there os some missing command, maybe there is a compatibility matrix somewhere.

**Reply by HighDoping** [reply]
Garnet [states](https://github.com/microsoft/garnet/blob/main/website/docs/welcome/compatibility.md) it's not a drop-in replacement for Redis. The [API compatibility page](https://github.com/microsoft/garnet/blob/main/website/docs/commands/api-compatibility.md) lists some missing APIs. It may need some further development to be compatiable.

**Comment by elf-pavlik** [comment]
An interesting benchmark comparing performance of Valkey and Redis: https://youtu.be/9hDvWVJtljE

**Comment by manast** [comment]
I don't know if it is well known, but Valkey is supported by BullMQ and we run the complete test suite in Valkey as part of our release process.
